enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,363 @@
+It will be updated from ``^([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$`` which accepts only regular DNS names to
+``^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$``.
+
+````go
+// EgressFirewallDestination is the endpoint that traffic is either allowed or denied to
+type EgressFirewallDestination struct {
+	// ..
+
+	// dnsName is the domain name to allow/deny traffic to. If this is set, cidrSelector must be unset.
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$
+	DNSName string `json:"dnsName,omitempty"`
+}
+````
+
+The following ``DNSName`` CRD will be added to the ``k8s.ovn.org`` api-group.
+
+````go
+// DNSName describes a DNS name used in a EgressFirewall rule.
+type DNSName struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior of DNSName.
+	Spec DNSNameSpec `json:"spec"`
+	// Observed status of DNSName
+	// +optional
+	Status DNSNameStatus `json:"status,omitempty"`
+}
+
+// DNSNameSpec is a desired state description of DNSName.
+type DNSNameSpec struct {
+	// IsRegular denotes whether the resource is representing a regular DNS name.
+	// Cannot be false if IsWildcard is also false.
+	// +optional
+	IsRegular bool `json:"isregular,omitempty"`
+	// IsWildcard denotes whether the resource is representing a wildcar DNS name.
+	// Cannot be false if IsRegular is also false.
+	// +optional
+	IsWildcard bool `json:"iswildcard,omitempty"`
+}
+
+type DNSNameStatus struct {
+	// The list of matching DNS names and their corresponding IPs along with TTL and next
+	// time to DNS lookup.
+	Items []DNSNameStatusItem `json:"items"`
+}
+
+type DNSNameStatusItem struct {
+	DNSName string        `json:"dnsname"`
+	Info    []DNSNameInfo `json:"info"`
+}
+
+type DNSNameInfo struct {
+	IP             string `json:"ip"`
+	TTL            string `json:"ttl"`
+	NextLookupTime string `json:"nextlookuptime"`
+}
+````
+
+### Implementation Details/Notes/Constraints [optional]
+
+The implementation changes needed for the proposed enhancement are documented in this section for each of the components.
+
+#### Cluster DNS Operator
+
+Cluster DNS Operator will deploy CoreDNS with the ``egressfirewall`` plugin enabled by adding it to the corefile. As the
+plugin will watch and update the ``DNSName`` CRs in the ``k8s.ovn.org`` api-group, proper RBAC permissions will be needed
+to be added to the ``ClusterRole`` for CoreDNS.
+
+#### CoreDNS
+
+The new plugin ``egressfirewall`` will be added to CoreDNS. As the plugin will inspect the DNS lookup queries and response from
+other plugins, it needs to be added before the other plugins (namely ``forward`` plugin) in the ``plugin.cfg`` file.
+
+The ``egressfirewall`` plugin will watch the ``DNSName`` CRs and whenever there is a DNS lookup which matches one of the ``DNSName``
+CRs (either regular or wildcard DNS names or both), then it will update the ``.status`` of the ``DNSName`` CR(s) if there's any change
```

Comment:
> what kind of informer, index, or data structure will the plugin use?

I am planning to use SharedIndexInformer. As for data structure, I am planning to use maps to store DNSName CR information. Do you have any suggestions?


> How is the matching done, especially with wildcards?

I have mentioned about this in this https://github.com/openshift/enhancements/pull/1335#discussion_r1157138722. I'll add the details in the EP.


> How do you mitigate performance impact when there are no matching CRs?

I have mentioned about this as well in this https://github.com/openshift/enhancements/pull/1335#discussion_r1157138722. I'll also add these details in the EP.


> Is the update done asynchronously? What happens if there is a conflict during an update?

Once the update to the DNSName CR is completed, then the response to the DNS lookup will be sent. If the update to the DNSName CR fails then the DNS lookup should also fail, otherwise this might cause discrepancy in the information between the pods and OVN-K. We can add retries to the DNSName CR update process.

Reply:
> As for data structure, I am planning to use maps to store DNSName CR information. Do you have any suggestions?

I was trying to think of a suitable data structure that could accommodate both regular and wildcard names/lookups, but using two maps as you suggested in https://github.com/openshift/enhancements/pull/1335#discussion_r1157138722 is probably better.